### PR TITLE
Refactoring the initial program

### PR DESCRIPTION
**Pull request by chipelmer**
*Added Product class
*Tried to implement Dapper

My database usage is rusty and the design/implementation strategy was developed in a silo...

### CODE REVIEW

File: RefactorKata/Program.cs [hunk 1]
```diff
@@ -1,41 +1,32 @@
...
 {
     class Program
     {
         static void Main(string[] args)
         {
-            //This is intentionally bad : (  Let's Refactor!
-            System.Data.SqlClient.SqlConnection Conn = new System.Data.SqlClient.SqlConnection("Server=.;Database=myDataBase;User Id=myUsername;Password = myPassword;");
+            // Not sure what this commented query indicates as far as the
+            // assignment is concerned or how the database is structured:
+            // cmd.CommandText = "SELECT * FROM Invoices";
```
Comment (codytwinton) [inline RefactorKata/Program.cs]: If you're not using code, remove it.

File: RefactorKata/Product.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                    Console.WriteLine("\n(Press any key to continue)");
+                    Console.ReadLine();
+                    return new List<Product>();
+                }
+            }
+        }
+
+        public static List<Product> ListAllWithoutUsingDapper()
+        {
+            var Conn = new SqlConnection(connectionString);
```
Comment (codytwinton) [inline RefactorKata/Product.cs]: Use camelCase for local variables.

File: RefactorKata/Product.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            var cmd = Conn.CreateCommand();
```
Comment (codytwinton) [inline RefactorKata/Product.cs]: Everything using the connection should be within the `using` statement.

File: RefactorKata/Product.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace RefactorKata
+{
+    public class Product
+    {
+        public string Name { get; set; }
+
+        private const string connectionString = "Server=.;Database=myDataBase;User Id=myUsername;Password=myPassword;Connection Timeout=2";
+
+        public static List<Product> ListAllUsingDapper()
```
Comment (codytwinton) [inline RefactorKata/Product.cs]: I'd change the return type to `IEnumerable<Product>`. More extensible.

File: RefactorKata/Product.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                {
+                    Console.WriteLine($"Error communicating with database:\n{e.Message}");
+                    Console.WriteLine("\n(Press any key to continue)");
+                    Console.ReadLine();
+                    return new List<Product>();
+                }
+            }
+        }
+
+        public static List<Product> ListAllWithoutUsingDapper()
```
Comment (codytwinton) [inline RefactorKata/Product.cs]: I'd change the return type to `IEnumerable<Product>`. More extensible.

File: RefactorKata/Program.cs (hunk 1 above)
Comment (chipelmer) [reply RefactorKata/Program.cs]: Removed

File: RefactorKata/Product.cs (hunk 2 above)
Comment (chipelmer) [reply RefactorKata/Product.cs]: Changed to camelCase

File: RefactorKata/Product.cs (hunk 3 above)
Comment (chipelmer) [reply RefactorKata/Product.cs]: Modified implementation of `using` statement to include all objects using it

File: RefactorKata/Product.cs (hunk 4 above)
Comment (chipelmer) [reply RefactorKata/Product.cs]: Return type changed

File: RefactorKata/Product.cs (hunk 5 above)
Comment (chipelmer) [reply RefactorKata/Product.cs]: Return type changed

**Comment by chipelmer** [comment]
New push is ready for review

File: RefactorKata/Product.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+        private const string connectionString = "Server=.;Database=myDataBase;User Id=myUsername;Password=myPassword;Connection Timeout=2";
+
+        public static IEnumerable<Product> ListAllUsingDapper()
+        {
+            using (var conn = new SqlConnection(connectionString))
+            {
+                try
+                {
+                    return conn.Query<Product>("SELECT Name FROM Products").ToList();
```
Comment (codytwinton) [inline RefactorKata/Product.cs]: `ToList()` is no longer necessary, since you are returning an `IEnumerable`

File: RefactorKata/Product.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                catch (Exception e)
+                {
+                    Console.WriteLine($"Error communicating with database:\n{e.Message}");
+                    Console.WriteLine("\n(Press any key to continue)");
+                    Console.ReadLine();
+                    return products;
+                }
+            }
+
+            return products;
```
Comment (codytwinton) [inline RefactorKata/Product.cs]: I'd move this `return products;` up to the try scope, or remove the `return products;` from the catch scope.

File: RefactorKata/Product.cs (hunk 6 above)
Comment (chipelmer) [reply RefactorKata/Product.cs]: Makes sense, but I'm not entirely sure I understand exactly what's going on with the return type being IEnumerable.  I'll have to read up on this...

File: RefactorKata/Product.cs (hunk 7 above)
Comment (chipelmer) [reply RefactorKata/Product.cs]: Good spot. Removed the `return` from the `catch` scope

**Pull request merged**